Document the abstract reader/writer interface

sjlongland opened the pull request:
`tinycbor` fork branch `thiagomacieira/dev` adds support for abstract readers and writers, but the implementation had some limitations in cases where multiple `CborValue` cursors were iterating over the same CBOR document, causing state contamination.  The interface also was not documented.

This documents the new interface and further builds upon it by slightly re-arranging `CborParser` members and opening the `token` in `CborValue` to use by the reader interface for any required purpose.

This pull request is now re-based on `intel/dev` and supercedes https://github.com/thiagomacieira/tinycbor/pull/2.

sjlongland committed:
cborparser: Document `cbor_parser_init_reader`.

Describe the input parameters for the function and how they are used as
best we understand from on-paper analysis of the C code.

sjlongland committed:
cborparser: Document `cbor_parser_init_reader`.

Describe the input parameters for the function and how they are used as
best we understand from on-paper analysis of the C code.

sjlongland committed:
cbor: Document the reader interface.

sjlongland committed:
cbor: Document the reader interface.

sjlongland committed:
cborparser: Pass CborValue to operation routines.

The `token` parameter is not sufficient since it is effectively shared
by all `CborValue` instances.  Since `tinycbor` often uses a temporary
`CborValue` context to perform some operation, we need to store our
context inside that `CborValue` so that we don't pollute the global
state of the reader.

sjlongland committed:
cborparser: Move `ops` outside of `union`

In its place, put an arbitrary `void *` pointer for reader context.  The
reader needs to store some context information which is specific to the
`CborParser` instance it is serving.  Right now, `CborValue::source::token`
serves this purpose, but the problem is that we also need a
per-`CborValue` context and have nowhere to put it.

Better to spend an extra pointer (4 bytes on 32-bit platforms) in the
`CborParser` (which there'll be just one of), then to do it in the
`CborValue` (which there may be several of) or to use a `CborReader`
object that itself carries two pointers (`ops` and the context, thus
we'd need an extra 3 pointers).

sjlongland committed:
cborparser: Move the reader context to CborParser.

sjlongland committed:
cborparser: Update documentation

sjlongland committed:
reader unit tests: Simplify the example reader

We simplify this reader in two ways:
1. we remove the `consumed` member of `struct Input`, and instead use
   the `CborValue`'s `source.token` member, which we treat as an
   unsigned integer offset into our `QByteArray`.
2. we replace the reader-specific `struct Input` with the `QByteArray`
   it was wrapping, since that's the only thing now contained in our
   `struct Input`.

If a `CborValue` gets cloned, the pointer referred to by `source.token`
similarly gets cloned, thus when we advance the pointer on the clone, it
leaves the original alone, so computing the length of unknown-length
entities in the CBOR document can be done safely.

sjlongland committed:
cborencoder: Document the write callback function.

What is not known, is what the significance is of
`CborEncoderAppendType`.  It basically tells the writer the nature of
the data being written, but the default implementation ignores this and
just blindly appends it no matter what.

That raises the question of why it's important enough that the writer
function needs to know about it.

sjlongland committed:
cborencoder: Document the write callback function.

What is not known, is what the significance is of
`CborEncoderAppendType`.  It basically tells the writer the nature of
the data being written, but the default implementation ignores this and
just blindly appends it no matter what.

That raises the question of why it's important enough that the writer
function needs to know about it.

sjlongland committed:
examples: Add buffered writer example.

sjlongland committed:
examples: Add buffered reader example

This reads a CBOR file piece-wise, seeking backward and forward through
the file if needed.  Some seeking can be avoided by tuning the block
size used in reads so that the read window shifts by smaller amounts.

sjlongland commented:
Forced commits to:
1. remove stale commit from the old `thiagomacieira/dev`
2. fix authorship of commits so that Github's PGP signature verification is happy.

thiagomacieira commented on src/cbor.h:
```diff
@@ -212,7 +212,20 @@ typedef enum CborEncoderAppendType
 } CborEncoderAppendType;
 
-typedef CborError (*CborEncoderWriteFunction)(void *, const void *, size_t, CborEncoderAppendType);
+/**
```
I prefer the documentation in the .c source file so the header is clean and small. Makes it easier to search for and find things in it, at a glance. Repeats.

sjlongland replied:
Ahh okay, I normally put docs in the headers out of habit because of packaging: when you ship a binary build, usually you ship a library object (`.a` / `.so` / `.dll`) and headers that describe the functions within; such users can't see comments in the C sources.

`tinycbor` being statically linked into the sources though and _not_ shipped separately, is a bit different.  I'll move the documentation comments. :-)

thiagomacieira replied:
Yeah, conflicting requirements. I prefer the comments in the sources because I usually read the .h itself. Of course, a good IDE can just collapse the comments...

In the olden days, that meant longer compile times for bigger source.

thiagomacieira commented on src/cbor.h:
```diff
@@ -390,8 +390,9 @@ struct CborParser
         const uint8_t *end;
-        const struct CborParserOperations *ops;
-    } source;
+        void *ctx;
```
Should we add a `cbor_parser_get_user_context()` so the user doesn't need to know about the internals of the struct? Then we don't have to document it either.

sjlongland replied:
Sounds fair, I'll add some "standard" getters/setters for structure hiding purposes.

thiagomacieira replied:
Like I said, we can do this after we merge your change.

thiagomacieira commented on tests/parser/tst_parser.cpp:
```diff
@@ -757,32 +757,32 @@ void tst_Parser::mapsAndArrays()
-        auto input = static_cast<Input *>(value->parser->data.ctx);
-        return input->data.size() - input->consumed >= int(len);
+        auto data = static_cast<QByteArray *>(value->parser->data.ctx);
+        auto consumed = uintptr_t(value->source.token);
```
I'd prefer the `uintptr_t` on the left side, but won't insist on your changing again.

sjlongland replied:
Yep, I've replaced the `auto` references.

thiagomacieira commented on tests/parser/tst_parser.cpp:
```diff
@@ -757,32 +757,32 @@ void tst_Parser::mapsAndArrays()
-        auto input = static_cast<Input *>(value->parser->data.ctx);
-        input->consumed += int(len);
+        auto consumed = uintptr_t(value->source.token);
+        consumed += int(len);
```
`int` is now wrong. Repeats.

sjlongland replied:
Replaced with `uintptr_t(len)`

thiagomacieira commented on tests/parser/tst_parser.cpp:
```diff
@@ -757,32 +757,32 @@ void tst_Parser::mapsAndArrays()
-        input->consumed += int(len);
+        auto consumed = uintptr_t(value->source.token);
+        consumed += int(len);
+        value->source.token = (void*)consumed;
```
Please write `reinterpret_cast<void *>(consumed)`

sjlongland replied:
Fixed, and also did the same for the cast done in `transfer_string` as well.

thiagomacieira reviewed (changes requested):
Hi Stuart

Thank you for your patience. This is looking really good. I will not insist on coding style changes in the examples (they're your code) nor on the `auto` usage. The `reinterpret_cast` is a different case, please update that and please move the Doxygen docs to the .c sources (See https://www.doxygen.nl/manual/docblocks.html#structuralcommands).

I will accept the change without the getter-setter functions working on tokens and context. We can discuss which ones to have after this is merged. I'd rather have wrapper functions so the users don't need to know about the internals of the object, so we can later change them if necessary, like you're doing now.

Finally, what's WSHUB-458? Sounds like a JIRA task number. Can you add a link to the commit message body to what this is and remove from the first line?

sjlongland commented:
> Thank you for your patience. This is looking really good. I will not insist on coding style changes in the examples (they're your code) nor on the `auto` usage. The `reinterpret_cast` is a different case, please update that and please move the Doxygen docs to the .c sources (See https://www.doxygen.nl/manual/docblocks.html#structuralcommands).

Ahh, the `auto` and `reinterpret_cast` are copied and pasted from the existing test cases.  I normally avoid such things but was trying to match the style of the existing test cases.  I will fix that. :-)

> Finally, what's WSHUB-458? Sounds like a JIRA task number. Can you add a link to the commit message body to what this is and remove from the first line?

Good point, force of habit due to the fact I was doing this at work.  I'll strip that token from the commit messages.

sjlongland committed:
cbor.h, cborencoder.c: Migrate documentation for encoder functions

sjlongland committed:
cbor.h, cborencoder.c: Migrate documentation for encoder functions

thiagomacieira commented:
(oops, accidentally edited instead of quoting)

> Ahh, the `auto` and `reinterpret_cast` are copied and pasted from the existing test cases. I normally avoid such things but was trying to match the style of the existing test cases. I will fix that. :-)

I'm using the Qt rules for `auto`: it has to be clear what it is, for someone reviewing on a dumb tool like GitHub. That usually means obvious (like iterators) or the type is visible on the same line.

sjlongland committed:
cbor.h, cborparser.c: Migrate parser documentation.

Not 100% sure of the syntax for documenting struct-members outside of
the struct as I'm too used to doing it inline, hopefully this works as
expected. :-)

sjlongland committed:
cbor.h, cborparser.c: Migrate parser documentation.

Not 100% sure of the syntax for documenting struct-members outside of
the struct as I'm too used to doing it inline, hopefully this works as
expected. :-)

sjlongland committed:
parser unit tests: Do not use `auto`, use reinterpret_cast

sjlongland commented:
> I'm using the Qt rules for `auto`: it has to be clear what it is, for someone reviewing on a dumb tool like GitHub. That usually means obvious (like iterators) or the type is visible on the same line.

Agreed, code clarity must be paramount.  I do more C than C++ so not used to using C++ features like `auto` and tend to avoid using it for that reason.

thiagomacieira reviewed (approved):
Let's merge and let me see about using this in Qt too.

thiagomacieira commented:
Gah, Travis hasn't run yet... Need to find someone to write GitHub Actions workflow, though likely that'll be me.

sjlongland commented:
Ahh `travis.org` is no more… I hit this on the weekend with one of my own projects (`aioax25`) and it's a landmine waiting to go blam for some of the @vrtsystems and @widesky projects (`pyat`, `cachefs`, `hszinc`) that are still set up to do CI on Travis-CI.  (And sadly, Github Actions doesn't hold a candle to Travis-CI, but Travis-CI is expensive now.)

thiagomacieira commented:
> Ahh `travis.org` is no more… I hit this on the weekend with one of my own projects (`aioax25`) and it's a landmine waiting to go blam for some of the @vrtsystems and @widesky projects (`pyat`, `cachefs`, `hszinc`) that are still set up to do CI on Travis-CI. (And sadly, Github Actions doesn't hold a candle to Travis-CI, but Travis-CI is expensive now.)

Too bad. Thanks for the update. Let me do a manual check locally and then force the merging.

sjlongland commented:
So, last time we looked at this, we got caught out by Travis CI's shutdown.  I've just done a rebase on the current `dev` branch to poke CI again.

sjlongland commented:
Second rebase to pick up the changes on `main` that weren't merged to `dev`.

sjlongland commented:
Okay, after picking up the changes in `main`, we have a merge conflict, so final rebase onto `dev` again to resolve the merge conflict.  This should unify the currently diverging `main` and `dev` branches once more.

thiagomacieira commented on examples/bufferedreader.c:
```diff
@@ -0,0 +1,783 @@
+/* vim: set sw=4 ts=4 et tw=78: */
```
Please add a header to each of the files with the licence text. Copy from the existing sources (not the existing example, that was my bad and I'll fix it). The copyright is **yours** not Intel.

sjlongland replied:
I was happy enough to contribute the copyright to Intel, but yeah I can put my name there.

sjlongland replied:
Just did this in a local branch (as well as `bufferedwriter.c`), I'll be pushing this shortly as a separate PR.

thiagomacieira replied:
> I was happy enough to contribute the copyright to Intel, but yeah I can put my name there.

That's unfortunately more complicated than it seems. Copyright assignment requires paperwork... ask anyone who has contributed to GNU software, which the FSF requires paperwork for.

sjlongland replied:
* [bufferedreader.c](https://github.com/intel/tinycbor/pull/326/commits/f98f354de2a1ea088410f0ea6bbc50e7101b01b8)
* [bufferedwriter.c](https://github.com/intel/tinycbor/pull/327/commits/34042b9ba98b3b601fe5badd7d4396be222224e8)

thiagomacieira commented on src/cbor.h:
```diff
@@ -321,11 +321,80 @@ enum CborParserIteratorFlags
+ * Defines an interface for abstract document readers.  This structure is used
+ * in conjunction with \ref cbor_parser_init_reader to define how the various
+ * required operations are to be implemented.
+ */
```
I prefer to keep the docs in the .c source, so the header is neat and small. But because we're documenting the members, which is annoying to do elsewhere, I'll allow it here.

thiagomacieira commented on src/cbor.h:
```diff
@@ -321,11 +321,80 @@ enum CborParserIteratorFlags
+    /**
+     * Overwrite the user-supplied pointer \a userptr with the address where the
+     * data indicated by \a offset is located, then advance the read pointer
+     * \a len bytes beyond that point.
```
Add that something like:
```
    This function should return \c CborNoError if there \a len bytes available at \a offset.
    It should return \c CborErrorUnexpectedEOF if there is not enough data. Other error
    conditions will be passed back to the user (e.g., \c CborErrorIO).
```

thiagomacieira commented on src/cbor.h:
```diff
@@ -334,25 +334,21 @@ struct CborParserOperations
      * \retval  false   Insufficient data is available to be read at this time.
      */
-    bool (*can_read_bytes)(void *token, size_t len);
+    bool (*can_read_bytes)(const struct CborValue *value, size_t len);
```
I want to discuss this change. What use-cases do you have in mind?

sjlongland replied:
Good question… I'll have to re-read the context because it was 2021 when I wrote this.

sjlongland replied:
Okay, found the commit where that change was made.  136321a03b37068e9a852eb937cc8934114f65df

`CborValue` objects represent where a given read is taking place to read *that* specific `CborValue`.  This allows for tracking of positions within files at a per-value point of view.

In the example buffered reader, it allocates a block of memory that acts as a "window" into the full file.  If a read happens that needs data that's within the window, the reader just `memcpy`'s it out.  Otherwise, if it's outside that window, the reader can attempt to slide that window back or forth as required.

The alternative would be for it to _always_ reload that buffer with content around the pointer of interest, which would be quite I/O intensive.  (My use case was reading/writing files out of flash filesystems like SPIFFS and littleFS.)

sjlongland replied:
This code block has moved to https://github.com/intel/tinycbor/pull/326

thiagomacieira commented on src/cborinternal_p.h:
```diff
@@ -205,15 +205,15 @@ static inline bool can_read_bytes(const CborValue *it, size_t n)
             return CBOR_PARSER_CAN_READ_BYTES_FUNCTION(it, n);
 #else
-            return it->parser->source.ops->can_read_bytes(it, n);
+            return it->parser->ops->can_read_bytes(it, n);
```
Then we could remove the `CborParserFlag_ExternalSource` flag and just use `it->parser->ops` as a check that it is external, no?

sjlongland replied:
That might be an option, not sure if the check for the external source was existing and I kept it there, or if I added it, but `it->parser->ops` should be `NULL` if it's internal.

thiagomacieira replied:
Don't change this now. We'll discuss the removal later, and possibly apply as a separate PR. My comment was just for us to start thinking.

sjlongland replied:
Change made in `496f0fb` (realised I should probably use `CBOR_NULLPTR`).

thiagomacieira commented on src/cborparser.c:
```diff
@@ -345,7 +345,7 @@ uint64_t _cbor_value_decode_int64_internal(const CborValue *value)
 {
     memset(parser, 0, sizeof(*parser));
-    parser->source.end = buffer + size;
+    parser->data.end = buffer + size;
```
Add:
```c
    parser->ops = CBOR_NULLPTR;
```

sjlongland replied:
Will do, I thought the `memset` would effectively do that, but being explicit never hurts.

thiagomacieira replied:
> Will do, I thought the `memset` would effectively do that, but being explicit never hurts.

It does for every architecture I am aware of, but it's not required. A null pointer does not have to be represented in memory by a pattern of zero bits. Since TinyCBOR is meant to be cross-platform to unusual microcontrollers, we should be thorough.

For everyone else, the compiler should detect a dead store and eliminate. See https://gcc.godbolt.org/z/o757Thdnd

sjlongland replied:
Taken care of in `4b152eb`.

thiagomacieira commented on tests/parser/tst_parser.cpp:
```diff
@@ -803,32 +803,32 @@ void tst_Parser::mapsAndArrays()
-        auto input = static_cast<Input *>(value->parser->data.ctx);
-        input->consumed += int(len);
+        auto consumed = uintptr_t(value->source.token);
+        consumed += int(len);
```
Nitpick: use `qsizetype` or `ptrdiff_t` here, to avoid truncation to 32 bits. I don't think we have a test that operates on more than 2 GB of RAM, but we could add one in the future. Repeats below.

sjlongland replied:
I note at some point it was changed to `uintptr_t`, but I've just adjusted it to `ptrdiff_t` in `0ec01d2`.

thiagomacieira commented on src/cbor.h:
```diff
@@ -216,7 +216,20 @@ typedef enum CborEncoderAppendType
+ *
+ * The \a append parameter informs the writer function whether it is writing
+ * a string or general CBOR data.
+ */
```
This one please move to the .c
```c
/**
 * \typedef CborEncoderWriteFunction
...
```

sjlongland replied:
Looks like this was done in commit 70dce85d.

thiagomacieira reviewed (changes requested):
Thank you, I have reviewed it. Some of the comments are stale because GitHub won't allow me to go back and edit/discard some of them (I'll dismiss after posting this). GitHub is not a nice review tool.

I'd like you to split this PR in three:
1. The pure doc changes for the interfaces as they currently exist, directly in the .c files (I've confirmed the `\var` works in Doxygen)
2. The CborParser changes
3. The CborEncoder changes

For the latter two, we need to write a good motivation for doing this.

sjlongland committed:
cbor.h: Remove empty whitespace

These crept in after documentation was added then subsequently removed.
